SWMR support

radistmorse opened the pull request:
This is the feature that was introduced in version 1.10 and allows to simultaneously read and write the same file. For this feature to work, you need to call `H5Fstart_swmr_write` on the writer side, and open the file with the flag `H5F_ACC_SWMR_READ` on the reader side. There are also some specific requirements, like all the attributes and groups need to be closed.

The `H5*close` functions are basically glorified destructors. Ultimately they call the same dec_ref as the highfive object destructor calls. But I decided to implement the explicit `close()` function for datasets, groups and attributes, in case user wants to start SWMR write before the relevant objects go out of scope.

Other functions that are needed for SWMR to work are `flush` and `refresh`  for the DataSets which were also added.

I added two new examples to show how SWMR is supposed to work. I didn't add any unit tests, as I'm not sure what exactly needs to be tested.

radistmorse commented:
I'll try to work something out of the shell script and split the functions in a separate PR. As for `close` - you're the boss here :) I described my motivation: the HDF5 documentation says that you need to explicitly `close` all unneeded objects before starting SWMR, and user might want to be able to do that, even though HighFive already does something similar under the hood. If you think you still don't want them, I'll remove them no problem, but then maybe some documentation updates would be nice, explaining how exactly one makes sure all the objects are closed before moving to SWMR.

1uc commented:
Sounds like your code doesn't absolutely need `close`. In that case, let's leave that change out for now.

To update the documentation, it might be best to add it directly to the examples. Or I can add a general section about how to close HighFive objects early.

In addition to std::optional there's also the Rust trick which doesn't require C++17:

```
void drop(DataSet&&) {}

int main() {
  auto dset = ...;
  drop(std::move(dset));
  // It's gone.
}
```

radistmorse commented:
I updated the PR. Now it only contains the SWMR functions. The examples won't work without merging the #69 first.

I also updated the bash script and added a timeout to the reader, so it should probably mostly work.

1uc commented:
Excellent, I've made small adjustments for some pedantic things. Feel free to squash them into your own commits.

Later, I'd like to check the opening mode. I thought we could have `AccessMode::SWMR` and then do `... | AccessMode::SWMR`; but I could easily be mistaken, I'll check later.

radistmorse commented on include/highfive/H5File.hpp:
```diff
@@ -122,6 +128,13 @@ class File: public Object, public NodeTraits<File>, public AnnotateTraits<File>
+#if H5_VERSION_GE(1, 10, 0)
+    /// \brief Starts SWMR write
+    ///
+    /// \see https://docs.hdfgroup.org/archive/support/HDF5/Tutor/swmr.html
```
hm, maybe it's not needed. If the user calls this function we could expect that he knows what are the requirements and limitation of SWMR mode. Let's assume users do not call the functions in random because their name look fancy.

radistmorse commented on include/highfive/bits/H5File_misc.hpp:
```diff
@@ -52,6 +61,9 @@ inline File::File(const std::string& filename,
     unsigned openMode = openFlags & (H5F_ACC_RDWR | H5F_ACC_RDONLY);
+#if H5_VERSION_GE(1, 10, 0)
+    openMode |= openFlags & (H5F_ACC_SWMR_READ | H5F_ACC_SWMR_WRITE);
+#endif
```
yes, if you want to open file for SWMR read/write you need to explicitly state it 
`H5Fopen (filename, H5F_ACC_RDONLY | H5F_ACC_SWMR_READ, H5P_DEFAULT);`
This is the whole point of these flags. Without them SWMR won't work.
